// @celo-tools/celo-ethers-wrapper/build/main/lib/CeloWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CeloWallet = void 0;
const ethers_1 = require("ethers");
const transactions_1 = require("./transactions");
const logger = new ethers_1.utils.Logger("CeloWallet");
const forwardErrors = [
    ethers_1.utils.Logger.errors.INSUFFICIENT_FUNDS,
    ethers_1.utils.Logger.errors.NONCE_EXPIRED,
    ethers_1.utils.Logger.errors.REPLACEMENT_UNDERPRICED,
];
class CeloWallet extends ethers_1.Wallet {
    /**
     * Override to skip checkTransaction step which rejects Celo tx properties
     * https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-signer/src.ts/index.ts
     */
    async populateTransaction(transaction) {
        const tx = await ethers_1.utils.resolveProperties(transaction);
        if (tx.to != null) {
            tx.to = Promise.resolve(tx.to).then((to) => this.resolveName(to));
        }
        if (tx.gasPrice == null) {
            tx.gasPrice = this.getGasPrice();
        }
        if (tx.nonce == null) {
            tx.nonce = this.getTransactionCount("pending");
        }
        if (tx.gasLimit == null) {
            tx.gasLimit = this.estimateGas(tx).catch((error) => {
                if (forwardErrors.indexOf(error.code) >= 0) {
                    throw error;
                }
                return logger.throwError("cannot estimate gas; transaction may fail or may require manual gas limit", ethers_1.utils.Logger.errors.UNPREDICTABLE_GAS_LIMIT, {
                    error: error,
                    tx: tx,
                });
            });
        }
        if (tx.chainId == null) {
            tx.chainId = this.getChainId();
        }
        else {
            tx.chainId = Promise.all([
                Promise.resolve(tx.chainId),
                this.getChainId(),
            ]).then((results) => {
                if (results[1] !== 0 && results[0] !== results[1]) {
                    logger.throwArgumentError("chainId address mismatch", "transaction", transaction);
                }
                return results[0];
            });
        }
        return ethers_1.utils.resolveProperties(tx);
    }
    /**
     * Override to serialize transaction using custom serialize method
     * https://github.com/ethers-io/ethers.js/blob/master/packages/wallet/src.ts/index.ts
     */
    async signTransaction(transaction) {
        const tx = await this.populateTransaction(transaction);
        if (tx.from != null) {
            if (ethers_1.utils.getAddress(tx.from) !== this.address) {
                logger.throwArgumentError("transaction from address mismatch", "transaction.from", transaction.from);
            }
            delete tx.from;
        }
        const signature = this._signingKey().signDigest(ethers_1.utils.keccak256((0, transactions_1.serializeCeloTransaction)(tx)));
        const serialized = (0, transactions_1.serializeCeloTransaction)(tx, signature);
        return serialized;
    }
    /**
     * Override just for type fix
     * https://github.com/ethers-io/ethers.js/blob/master/packages/wallet/src.ts/index.ts
     */
    sendTransaction(transaction) {
        return super.sendTransaction(transaction);
    }
    /**
     * Override to skip checkTransaction step which rejects Celo tx properties
     * https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-signer/src.ts/index.ts
     */
    async estimateGas(transaction) {
        this._checkProvider("estimateGas");
        const tx = await ethers_1.utils.resolveProperties(transaction);
        return await this.provider.estimateGas(tx);
    }
    /**
     * Override to support alternative gas currencies
     * https://github.com/celo-tools/ethers.js/blob/master/packages/abstract-signer/src.ts/index.ts
     */
    async getGasPrice(feeCurrencyAddress) {
        this._checkProvider("getGasPrice");
        // @ts-ignore
        return await this.provider.getGasPrice(feeCurrencyAddress);
    }
}
exports.CeloWallet = CeloWallet;
